workflow/templateresolution/context.go

```diff
@@ -14,19 +16,60 @@ import (
 	"github.com/argoproj/argo-workflows/v3/workflow/common"
 )
 
+var (
+	defaultCacheTTL = time.Minute
+)
+
+type workflowTemplateCache struct {
+	createdAt time.Time
+	template  *wfv1.WorkflowTemplate
+}
+
+var _ WorkflowTemplateNamespacedGetter = (*workflowTemplateInterfaceWrapper)(nil)
+
 // workflowTemplateInterfaceWrapper is an internal struct to wrap clientset.
 type workflowTemplateInterfaceWrapper struct {
 	clientset typed.WorkflowTemplateInterface
+	cache     map[string]workflowTemplateCache
+	cacheMu   sync.Mutex
+	cacheTTL  time.Duration
+}
+
+func WrapWorkflowTemplateInterface(clientset typed.WorkflowTemplateInterface) *workflowTemplateInterfaceWrapper {
+	return &workflowTemplateInterfaceWrapper{
+		clientset: clientset,
+		cache:     map[string]workflowTemplateCache{},
+		cacheTTL:  defaultCacheTTL,
+	}
 }
 
-func WrapWorkflowTemplateInterface(clientset typed.WorkflowTemplateInterface) WorkflowTemplateNamespacedGetter {
-	return &workflowTemplateInterfaceWrapper{clientset: clientset}
+func (w *workflowTemplateInterfaceWrapper) WithCacheTTL(ttl time.Duration) *workflowTemplateInterfaceWrapper {
+	w.cacheTTL = ttl.Abs()
+	return w
 }
 
 // Get retrieves the WorkflowTemplate of a given name.
 func (wrapper *workflowTemplateInterfaceWrapper) Get(name string) (*wfv1.WorkflowTemplate, error) {
 	ctx := context.TODO()
-	return wrapper.clientset.Get(ctx, name, metav1.GetOptions{})
+
+	wrapper.cacheMu.Lock()
+	defer wrapper.cacheMu.Unlock()
+
+	val, ok := wrapper.cache[name]
```

Comment:
rather than a manual cache, this could probably accept a cache as an argument.

for instance, the Controller already has a cache of all WorkflowTemplates in its Informer.
the Server does not have an Informer yet, but will likely eventually have a SQLite cache similar to the one for the Workflow List
the CLI generally has none of the above though

Reply:
It makes sense - I did this as a quick resolution.
Do you prefer to use the same pattern as with workflow list or just add template informer to the server?

Reply:
#13672 is possibly fine for now, but the SQLite cache will eventually be needed for other features for filtering for WorkflowTemplates. It doesn't seem to be an often used part of the UI though, so maybe it's fine to leave without for now and modify when needed.

Reply:
FYI: #13672 is ready for review!